swagger_server/__main__.py

```diff
@@ -34,6 +35,10 @@ def start_consumer(thread_queue, db_instance):
     t1.start()
 
 
+def start_pull_topology_change():
+    call(["python", "swagger_server/jobs/pull_topo_changes.py"])
```

Comment:
Wouldn't it be simpler if this is a function call rather than creating a subprocess?  If it has to be a subprocess, it would be useful to have the reasons documented here.

Reply:
This is a subprocess because the pull_topo_changes.py has its own main. Added a comment.

Reply:
I am not sure that alone is a good reason though. :-)

Calling a function is simpler than creating a subprocess.  For example, with a subprocess, what happens when the parent or child exits or crashes or gets killed?  Will we be able to correctly clean up everything?  Will there be zombie processes?  A subprocess is also costlier, since we will have a whole new Python interpreter and code taking up memory and CPU and other resources.

It is okay for the puller to have a main method, for testing and the such.  But if possible, we should make this a function call.

Reply:
The pull_topo_change should be a separate process from the sdx-lc, that is constantly running. This would make the code structure clearer as well. Even if the parent gets killed, it should still keep pulling topology from domain controller, and publish to sdx-controller if changes are found.

Reply:
That makes sense. It would be good to have this intention documented in a comment (so that future readers, probably ourselves, will know this), and have this assumption (that child will continue running, and without races) tested.

However, if the parent exits, and then gets restarted (by someone or something, such as a monitoring system), wouldn't we be starting an extra pull process?

Reply:
Sounds good. Added a comment. 
Yes that way we will start an extra process. But for most of the cases, if something goes wrong, we will restart the container, which will restart both the sdx-lc and topology-pulling.

Reply:
Hmm. In that case the puller might just as well be a thread rather than a process?

Reply:
For threads, they will die when all your non-daemon threads (e.g. the main thread) have exited. So in that case, if sdx-lc dies, the pull-topo thread will also stop. We want to keep pull-topo running independently.

Reply:
If I understand correctly, the Python process will exit when all threads have exited or joined.  It will keep running even if the main thread crashed.

If we're going to restart the container when sdx-lc dies anyway, pull-topo also will restart along with that -- so it doesn't matter if it is a separate process, right?

It is just that a thread seems "cleaner" to me in this instance. I could be wrong though. :-)

Reply:
Yeah it's nice to make sure we are on the right track.
I think the topo-pulling, as a "job", it should be independent. The sdx-lc will only responsible for starting the topo-pulling.
Usually once the main thread exits, the whole Python process will exit, assuming there are no other non-daemon threads running. 
https://superfastpython.com/main-thread/#:~:text=The%20main%20thread%20in%20each,other%20non%2Ddaemon%20threads%20running.

@YufengXin any ideas?

Reply:
I remember having to explicitly stop the keep-alive thread in sdx-controller.

https://github.com/atlanticwave-sdx/sdx-controller/pull/94/commits/fdc3588824dc67727cdf23a628442319386838c1

Before adding that, test suite would not quit (after tests are run) without hitting Control-C, which is not an option when running tests on CI.

Reply:
I did a little experiment:

https://gist.github.com/sajith/f0b856080da0654be6fc364c835fa4f1

The subthread will continue on even if the main thread dies.  The whole process exits when the subthread also dies.

Reply:
> I remember having to explicitly stop the keep-alive thread in sdx-controller.
> 
> [atlanticwave-sdx/sdx-controller@fdc3588](https://github.com/atlanticwave-sdx/sdx-controller/commit/fdc3588824dc67727cdf23a628442319386838c1)
> 
> Before adding that, test suite would not quit (after tests are run) without hitting Control-C, which is not an option when running tests on CI.

They are different. The sdx-controller example, has the rabbitmq thread in the swagger_server process:
```
root@a6ec8e449462:/usr/src/app# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1 99.4  0.4 860716 79344 pts/0    Rsl+ 15:18   2:20 python3 -m swagger_server
```
But the pull_topo subprocess lives outside swagger_server process:
```
root@a891db8c88d0:/usr/src/app# ps aux
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root           1  0.5  0.5 718584 46428 pts/0    Ssl+ 15:22   0:00 python3 -m swagger_server
root          11  0.3  0.3 627532 31324 pts/0    Sl+  15:22   0:00 python swagger_server/jobs/pull_topo_changes.py
```

Reply:
> I did a little experiment:
> 
> https://gist.github.com/sajith/f0b856080da0654be6fc364c835fa4f1
> 
> The subthread will continue on even if the main thread dies. The whole process exits when the subthread also dies.

Yes, this should be exactly what we need.